dspy/adapters/two_step_adapter.py

```diff
@@ -0,0 +1,167 @@
+            inputs: The current input
+        
+        Returns:
+            A list of messages to be passed to the main LM.
+        """
+        messages = []
+
+        # Create a task description for the main LM
+        task_description = self._create_task_description(signature)
+        messages.append({"role": "system", "content": task_description})
+
+        # Format demos in a natural way
+        for demo in demos:
+            messages.extend(self._format_demo(signature, demo))
+
+        # Format the current input
+        messages.append({"role": "user", "content": self._format_input(signature, inputs)})
+
+        return messages
+
+    def parse(self, signature: Signature, completion: str) -> dict[str, Any]:
+        """
+        Use a smaller LM (extraction_model) with chat adapter to extract structured data 
+        from the raw completion text of the main LM.
+
+        Args:
+            signature: The signature of the original task
+            completion: The completion from the main LM
+
+        Returns:
+            A dictionary containing the extracted structured data.
+        """
+        # The signature is supposed to be "text -> {original output fields}"
+        extractor_signature = self._create_signature_with_text_input_and_outputs(signature)
+        
+        try:
+            # Call the smaller LM to extract structured data from the raw completion text with ChatAdapter
+            parsed_result = ChatAdapter()(lm=self.extraction_model, lm_kwargs={}, signature=extractor_signature, demos=[], inputs={ "text": completion })
+            return parsed_result[0]
+
+        except Exception as e:
+            raise ValueError(f"Failed to parse response from the original completion: {completion}") from e
+    
+    def _create_task_description(self, signature: Signature) -> str:
+        """Create a natural description of the task based on the signature"""
+        parts = []
+        
+        parts.append("You are a helpful assistant that can solve tasks based on user input.")
+        parts.append("Your input fields are:\n" + enumerate_fields(signature.input_fields))
+        parts.append("Your final output fields are:\n" + enumerate_fields(signature.output_fields))
+        parts.append("You should provide reasoning for your answers in details so that your answer can be understood by another agent")
+        
+        if signature.instructions:
+            parts.append(f"Specific instructions: {signature.instructions}")
+            
+        return "\n".join(parts)
+    
+    def _format_input(self, signature: Signature, values: dict[str, Any]) -> str:
+        parts = []
+        
+        for name in signature.input_fields.keys():
+            if name in values:
+                parts.append(f"{name}: {values[name]}")
+                
+        return "\n".join(parts)
+    
+    def _format_demo(self, signature: Signature, values: dict[str, Any]) -> list[dict[str, str]]:
```

Comment:
I am curious about one thing - the few-shot example's output part (assistant message) could be different from the main LM's output, since we don't require the output format from the main LM. Could this be a problem? One way to verify is running optimizer on a simple CoT, and use this two-step adapter, then compare the main LM's output between the original and optimized CoT. 

My guess is this should be all right, since reasoning model is pretty strong at understanding user intention, so this slight mismatch won't affect the response quality.

Reply:
Great point, actually the main LM tries to output a text in a semi-structured way like the one below. The good point of this adapter is that it does not require reasoning models to strictly follow a specific format.

```
reasoning: First, we determine Ella's correct answers by subtracting her incorrect answers from the total: 40 − 4 = 36. Next, we compute half of Ella’s score, which is ½ × 36 = 18. Marion's score is stated as 6 more than half of Ella's score; thus, Marion’s score is 18 + 6 = 24. answer: 24
```